File: assemblyscript/wifi.ts

@external("wifi", "wifiStatus")
declare function _wifiStatus(): u32;

@external("wifi", "wifiConnect")
declare function _wifiConnect(ssid: usize, ssid_len: i32, pass: usize, pass_len: i32): void;

@external("wifi", "wifiLocalIp")
declare function _wifiLocalIp(ptr: usize): void;

export function connect(ssid: string, pass: string): void {
  let outSsid = String.UTF8.encode(ssid);
  let outPass = String.UTF8.encode(pass);
  _wifiConnect(
    changetype<usize>(outSsid), outSsid.byteLength,
    changetype<usize>(outPass), outPass.byteLength
  )
}

export function status(): u32 {
  return _wifiStatus();
}

export function localIp(): string {
  const ptr = memory.data(16);
  _wifiLocalIp(ptr);
  return String.UTF8.decodeUnsafe(ptr, 16, true);
}

export const WL_NO_SHIELD: u32 = 0x255;
export const WL_IDLE_STATUS: u32 = 0x0;
export const WL_NO_SSID_AVAIL: u32 = 0x1;
export const WL_SCAN_COMPLETED: u32 = 0x2;
export const WL_CONNECTED: u32 = 0x3;
export const WL_CONNECT_FAILED: u32 = 0x4;
export const WL_CONNECTION_LOST: u32 = 0x5;
export const WL_DISCONNECTED: u32 = 0x6;

File: tinygo/wifi/wifi.go
package wifi

//go:wasm-module wifi
//go:export wifiStatus
func Status() uint

//go:wasm-module wifi
//go:export wifiConnect
func Connect(ssid string, password string)

//go:wasm-module wifi
//go:export wifiLocalIp
func _getLocalIp(buf *byte)

func LocalIp() string {
	var buf = make([]byte, 16)
	_getLocalIp(&buf[0])
	// Find '\0'
	n := -1
	for i, b := range buf {
		if b == 0 {
			break
		}
		n = i
	}
	return string(buf[:n+1])
}

const (
	WL_NO_SHIELD       = 0x255
	WL_IDLE_STATUS     = 0x0
	WL_NO_SSID_AVAIL   = 0x1
	WL_SCAN_COMPLETED  = 0x2
	WL_CONNECTED       = 0x3
	WL_CONNECT_FAILED  = 0x4
	WL_CONNECTION_LOST = 0x5
	WL_DISCONNECTED    = 0x6
)
